src/transport.rs

```diff
@@ -182,6 +185,54 @@ impl Transport {
         req.map_err(Error::Hyper)
     }
 
+    /// Makes an HTTP request, upgrading the connection to a TCP
+    /// stream on success.
+    ///
+    /// This method can be used for operations such as viewing
+    /// docker container logs interactively.
+    pub fn stream_upgrade<B>(
+        &self,
+        method: Method,
+        endpoint: &str,
+        body: Option<(B, Mime)>,
+    ) -> impl Future<Item = impl AsyncRead + AsyncWrite, Error = Error>
+    where
+        B: Into<Body>
+    {
+        match self {
+            Transport::Tcp { .. } | Transport::EncryptedTcp { .. } => (),
+            _ => panic!("connection streaming is only supported over TCP"),
+        };
+
+        let req = self.build_request(method, endpoint, body, |builder| {
+            builder.header(header::CONNECTION, "Upgrade")
+                .header(header::UPGRADE, "tcp");
+        }).expect("Failed to build request!");
+
+        self.send_request(req).and_then(|res| {
+            if res.status() != StatusCode::SWITCHING_PROTOCOLS {
+                panic!("Our server didn't upgrade: {}", res.status());
```

Comment:
It might be better to return a failed Future here, instead of panicking?

Reply:
Good idea

Reply:
Fixed